### PR TITLE
[PINOT-3104]: StarTree: Enhancements and performance improvements.

### PR DESCRIPTION
**Pull request by mayankshriv**
1. Split leaf nodes based on time.
2. Add aggregated value on intermediate nodes.
3. Optimization for StarTreeIndexOperator to run remove group-by columns
from remaining predicates, after splitting on them.
Tests to follow in another PR.

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java [hunk 1]
```diff
@@ -49,84 +50,83 @@
...
   private static final Logger LOGGER = LoggerFactory.getLogger(StarTreeIndexOperator.class);
 
   private IndexSegment segment;
-  //predicates that can be applied on the star tree
-  Map<String, PredicateEntry> eligibleEqualityPredicatesMap;
-  //predicates that cannot be applied on star tree index because it appears in group by clause
-  Map<String, PredicateEntry> inEligiblePredicatesMap;
-  //predicates that cannot be applied because they are not of type Equality
-  Map<String, PredicateEntry> nonEqualityPredicatesMap;
+  /**
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Why use javadoc style comment here?

Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Resolved.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java [hunk 2]
```diff
@@ -49,84 +50,80 @@
...
     if (childFilter.getOperator() == FilterOperator.EQUALITY) {
       EqPredicate eqPredicate = (EqPredicate) predicate;
-      //computing dictionaryId allows us early termination and avoids multiple looks up during tree traversal
+      // Computing dictionaryId allows us early termination and avoids multiple looks up during tree
+      // traversal
       int dictId = dictionary.indexOf(eqPredicate.getEqualsValue());
       if (dictId < 0) {
-        //empty result
+        // Empty result
         emptyResult = true;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: When we set emptyResult to true, should we just return because all other operations are redundant?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java [hunk 3]
```diff
@@ -297,65 +355,118 @@ public BlockId getId() {
...
     while (!searchQueue.isEmpty()) {
       SearchEntry searchEntry = searchQueue.remove();
       StarTreeIndexNode current = searchEntry.starTreeIndexnode;
-      HashSet<String> remainingColumnsToFilter = searchEntry.remainingPredicates;
-      //check if its leaf
-      if (current.isLeaf()) {
-        //reached leaf
+      HashSet<String> remainingPredicateColumns = searchEntry.remainingPredicateColumns;
+      HashSet<String> remainingGroupByColumns = searchEntry.remainingGroupByColumns;
+      // Check if its leaf
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Update comments

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java [hunk 4]
```diff
@@ -297,65 +355,118 @@ public BlockId getId() {
...
+          newRemainingGroupByColumns);
+    }
+    return matchedEntries;
+  }
+
+  /**
+   * Helper method to add matching children into the search queue.
+   * - If predicate can be applied (i.e. equality predicate that is eligible), add the child
+   * satisfying
+   * the predicate into the queue.
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Move to previous line?

Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Append to the previous line

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java [hunk 5]
```diff
@@ -297,65 +355,118 @@ public BlockId getId() {
...
+      nextValueId = predicateEntry.dictionaryId;
+      remainingPredicateColumns.remove(column);
+      remainingGroupByColumns.remove(column);
+      if (children.containsKey(nextValueId)) {
+        addNodeToSearchQueue(searchQueue, children.get(nextValueId), remainingPredicateColumns,
+            remainingGroupByColumns);
+      }
+    } else {
+      int nextValueId;
+      if (groupByColumns.contains(column) || predicatesMap.containsKey(column)
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Will it be more intuitive using remainingGroupByColumns and remainingPredicateColumns?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java [hunk 6]
```diff
@@ -297,65 +355,118 @@ public BlockId getId() {
...
+          || !children.containsKey(StarTreeIndexNode.all())) {
+        for (StarTreeIndexNode indexNode : children.values()) {
+          if (indexNode.getDimensionValue() != StarTreeIndexNode.all()) {
+            remainingGroupByColumns.remove(column);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Also remove column from remainingPredicateColumns?

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java [hunk 7]
```diff
@@ -297,65 +355,118 @@ public BlockId getId() {
...
-            newEntry.remainingPredicates = newRemainingPredicates;
-            searchQueue.add(newEntry);
+      // Find next set of nodes to search
+      String nextDimension =
+          dimensionIndexToNameMapping.inverse().get(current.getChildDimensionName());
+
+      HashSet<String> newRemainingPredicateColumns = new HashSet<>();
+      newRemainingPredicateColumns.addAll(remainingPredicateColumns);
+      HashSet<String> newRemainingGroupByColumns = new HashSet<>();
+      newRemainingGroupByColumns.addAll(remainingGroupByColumns);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: No need to create new Sets because the searchEntry has been removed from searchQueue.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java [hunk 8]
```diff
@@ -297,65 +355,118 @@ public BlockId getId() {
...
+      }
+    } else {
+      int nextValueId;
+      if (groupByColumns.contains(column) || predicatesMap.containsKey(column)
+          || !children.containsKey(StarTreeIndexNode.all())) {
+        for (StarTreeIndexNode indexNode : children.values()) {
+          if (indexNode.getDimensionValue() != StarTreeIndexNode.all()) {
+            remainingGroupByColumns.remove(column);
+            addNodeToSearchQueue(searchQueue, indexNode, remainingPredicateColumns,
+                remainingGroupByColumns);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: For each child, create a new remainingPredicateColumns and a new remainingGroupByColumns, otherwise they are sharing the same instances.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java (hunk 2 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Will evaluate if that can cause an issue, and submit another PR if required.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java (hunk 7 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Need to create a copy for each node. Not doing so means nodes may have the reference to the same 'Set' which is being modified, and we don't want that.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java (hunk 4 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java (hunk 6 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: In the 'else' condition, it implies that the column was never in the PredicateColumns for the node. So no need to remove.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java (hunk 8 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Sharing is fine for children of the same parent. It makes the 'remove' redundant, but saves a lot of garbage.

**Comment by Jackie-Jiang** [comment]
LGTM.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java (hunk 3 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Addressed before pushing.

File: pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java (hunk 5 above)
Comment (mayankshriv) [inline pinot-core/src/main/java/com/linkedin/pinot/core/operator/filter/StarTreeIndexOperator.java]: Addressed before pushing.
